README.md
# rehash
Resumable hashlib: a picklable, reentrant interface to CPython's OpenSSL-based hashlib standard library


Changes.rst
Changes for v1.0.1 (2022-08-28)
===============================

-  Disclaim support for OpenSSL 3

-  Test and release infrastructure improvements

Changes for v1.0.0 (2019-10-23)
===============================

-  Update EVPobject layout for Python 3.8

-  Test and release infrastructure improvements

-  Bump version to 1.0.0 to indicate stable API

Changes for v0.3.1 (2017-11-23)
===============================

Fix EVPobject layout in Python 3.5.0 through 3.5.2

Changes for v0.3.0 (2017-08-08)
===============================

-  Support Python >= 2.7, < 2.7.9, which does not have
   hashlib.algorithms\_guaranteed

-  Don't expose unsupported hashers

-  Documentation improvements

Changes for v0.2.1 (2017-07-13)
===============================

-  Hashlib API compat fixes

Changes for v0.2.0 (2017-07-13)
===============================

-  Fix for version-dependent compat with EVPobject

Changes for v0.1.0 (2017-07-12)
===============================

-  Initial release



test/test.py
#!/usr/bin/env python

import os, sys, unittest, tempfile, json, logging, pickle, hashlib, ssl

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))) # noqa

import rehash

print("OpenSSL version:", ssl.OPENSSL_VERSION)

class TestResumableHasher(unittest.TestCase):
    def setUp(self):
        logging.basicConfig(level="DEBUG")

    def assert_resumable(self, hasher):
        hasher.update(b"foo")
        state = pickle.dumps(hasher)
        hasher2 = pickle.loads(state)
        hasher.update(b"bar"*1024*1024)
        hasher2.update(b"bar"*1024*1024)
        self.assertEqual(hasher.digest(), hasher2.digest())

    def test_basic_statements(self):
        for algorithm in hashlib.algorithms_guaranteed:
            if algorithm.startswith("blake2") or algorithm.startswith("sha3") or algorithm.startswith("shake"):
                with self.assertRaises(Exception):
                    rehash.ResumableHasher(algorithm.lower())
            else:
                print(algorithm)
                self.assert_resumable(rehash.new(algorithm.lower()))
                self.assert_resumable(rehash.new(algorithm.lower(), b"initial_data"))
                self.assert_resumable(rehash.ResumableHasher(algorithm.lower()))
                self.assert_resumable(rehash.ResumableHasher(algorithm.lower(), b"initial_data"))
                self.assert_resumable(getattr(rehash, algorithm)())
                self.assert_resumable(getattr(rehash, algorithm)(b"initial_data"))

    def test_repr(self):
        # Some of the memory at the tail of the hasher state is uninitialized at initial state
        self.assertEqual(
            repr(rehash.sha256())[:64],
            "rehash.sha256(state='eJxLf8aZ1boufXfR5zwbq6/+S+uD+AJ7Mlhnr77ZLC959kE0AxWBAhDfa7Ku5N63OAIAWHETSw==')"[:64]
        )
        rehash.opaque_repr = True
        self.assertEqual(
            repr(rehash.sha256()),
            "rehash.sha256()"
        )

    def test_doc_example(self):
        import pickle, rehash
        hasher = rehash.sha256(b"foo")
        state = pickle.dumps(hasher)

        hasher2 = pickle.loads(state)
        hasher2.update(b"bar")

        assert hasher2.hexdigest() == rehash.sha256(b"foobar").hexdigest()


if __name__ == '__main__':
    unittest.main()


rehash/__init__.py
import os, sys, hashlib, base64, zlib
from ctypes import cast, memmove, POINTER, c_void_p
from ssl import OPENSSL_VERSION

from .structs import EVPobject

opaque_repr = False

class ResumableHasher(object):
    name = None
    _algorithms_guaranteed = getattr(hashlib,
                                     "algorithms_guaranteed",
                                     ["md5", "sha1", "sha224", "sha256", "sha384", "sha512"])

    def __init__(self, name=None, data=None, state=None):
        # Starting with OpenSSL 3.0, the digest algorithm implementations use the "provider" interface described here:
        # https://www.openssl.org/docs/manmaster/man7/provider-digest.html
        # Support for serializing digest contexts for these implementations is not yet implemented.
        if OPENSSL_VERSION >= "OpenSSL 3.0.0":
            raise NotImplementedError("OpenSSL 3 is not yet supported. Please use OpenSSL 1.x")
        if state is not None:
            if not self.name:
                raise Exception('Parameter "name" is required')
            self.__setstate__(state=dict(name=name, md_data=zlib.decompress(base64.b64decode(state))))
            if data is not None:
                self.update(data)
            return
        if self.name is not None:
            data = name
        else:
            self.name = name
        if not self.name:
            raise Exception('Parameter "name" is required')
        hasher_args = [] if data is None else [data]
        self._hasher = self._get_hashlib_hasher(self.name)(*hasher_args)

    def _get_hashlib_hasher(self, name):
        if name.startswith("blake2"):
            raise Exception("blake2 algorithms are not OpenSSL-based and not supported by rehash")
        if name.startswith("sha3"):
            # FIXME: sha3 is supported by openssl 1.1.1+ and default in cpython
            raise Exception("sha3 algorithms are not supported by rehash")
        if name.startswith("shake"):
            # FIXME: shake is supported by openssl 1.1.1+ and default in cpython
            raise Exception("shake algorithms are not supported by rehash")
        if name in self._algorithms_guaranteed:
            return getattr(hashlib, name)
        else:
            return hashlib.new(name)

    def _get_evp_md_ctx(self):
        c_evp_obj = cast(c_void_p(id(self._hasher)), POINTER(EVPobject))
        print(type(c_evp_obj), c_evp_obj)
        if hasattr(c_evp_obj.contents.ctx, "contents"):
            return c_evp_obj.contents.ctx.contents
        else:
            return c_evp_obj.contents.ctx

    def __getstate__(self):
        ctx = self._get_evp_md_ctx()
        ctx_size = ctx.digest.contents.ctx_size
        hasher_state = ctx.md_data[:ctx_size]
        return dict(name=self.name, md_data=hasher_state)

    def __setstate__(self, state):
        self.name = state["name"]
        self._hasher = self._get_hashlib_hasher(self.name)()
        ctx = self._get_evp_md_ctx()
        ctx_size = ctx.digest.contents.ctx_size
        memmove(ctx.md_data, state["md_data"], ctx_size)

    def __getattr__(self, a):
        return getattr(self._hasher, a)

    def __repr__(self):
        if opaque_repr:
            return "{}.{}()".format(self.__module__, self.__class__.__name__)
        else:
            md_data = base64.b64encode(zlib.compress(self.__getstate__()["md_data"])).decode()
            return "{}.{}(state='{}')".format(self.__module__, self.name, md_data)


new = ResumableHasher

def _initialize():
    module = sys.modules[__name__]
    for name in ResumableHasher._algorithms_guaranteed:
        if name.startswith("blake2") or name.startswith("sha3") or name.startswith("shake"):
            continue
        setattr(module, name, type(name, (ResumableHasher,), dict(name=name)))


_initialize()


rehash/structs.py
from ssl import OPENSSL_VERSION
from sys import version_info as PYTHON_VERSION
from ctypes import c_void_p, POINTER, Structure, c_int, c_ulong, c_char, c_size_t, c_ssize_t, py_object

PyObject_HEAD = [
    ('ob_refcnt', c_size_t),
    ('ob_type', c_void_p)
]

# OpenSSL struct evp_md_st
# OpenSSL 1.0.2 and earlier:
# https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/evp/evp.h#L159-L181
# OpenSSL 1.1.0:
# https://github.com/openssl/openssl/blob/OpenSSL_1_1_0-stable/crypto/include/internal/evp_int.h#L88-L102
# OpenSSL 1.1.1 and later:
# https://github.com/openssl/openssl/blob/master/include/crypto/evp.h#L247-L288
class EVP_MD(Structure):
    _fields_ = [
        ('type', c_int),
        ('pkey_type', c_int),
        ('md_size', c_int),
        ('flags', c_ulong)
    ]
    if OPENSSL_VERSION >= "OpenSSL 3.0.0":
        _fields_ += [
            ('origin', c_int),
        ]
    _fields_ += [
        ('init', c_void_p),
        ('update', c_void_p),
        ('final', c_void_p),
        ('copy', c_void_p),
        ('cleanup', c_void_p),
    ]
    if OPENSSL_VERSION < "OpenSSL 1.1.0":
        _fields_ += [
            ('sign', c_void_p),
            ('verify', c_void_p),
            ('required_pkey_type', c_int * 5),
        ]
    _fields_ += [
        ('block_size', c_int),
        ('ctx_size', c_int),
    ]
    print("EVP_MD fields", _fields_)

# OpenSSL struct evp_md_ctx_st
# OpenSSL 1.1.0 and earlier:
# https://github.com/openssl/openssl/blob/master/crypto/evp/evp_locl.h#L12-L22
# OpenSSL 1.1.1 and later, before 3.0.0:
# https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/evp/evp_local.h#L12-L22
# OpenSSL 3.0.0 and later:
# https://github.com/openssl/openssl/blob/master/crypto/evp/evp_local.h#L16-L34
class EVP_MD_CTX(Structure):
    _fields_ = [
        ('digest', POINTER(EVP_MD)),
        ('engine', c_void_p),
        ('flags', c_ulong),
        ('md_data', POINTER(c_char)),
    ]
    if OPENSSL_VERSION >= "OpenSSL 3.0.0":
        _fields_ = [
            ('reqdigest', POINTER(EVP_MD)),
        ] + _fields_ + [
            ('pctx', c_void_p),
            ('update', c_void_p),
            ('algctx', c_void_p),
            ('fetched_digest', POINTER(EVP_MD)),
        ]

# Python 3.8+: https://github.com/python/cpython/blob/master/Modules/_hashopenssl.c#L58-L64
# Python 3.5.3 - 3.7: https://github.com/python/cpython/blob/3.7/Modules/_hashopenssl.c#L52-L59
# Python 3 - 3.5.2: https://github.com/python/cpython/blob/3.4/Modules/_hashopenssl.c#L39-L46
# Python 2.7.13+: https://github.com/python/cpython/blob/2.7/Modules/_hashopenssl.c#L71-L78
class EVPobject(Structure):
    _fields_ = PyObject_HEAD + (
        [("name", POINTER(py_object))] if PYTHON_VERSION < (3, 8) else []
    ) + [
        ("ctx", EVP_MD_CTX if (3, 0) < PYTHON_VERSION < (3, 5, 3) or PYTHON_VERSION < (2, 7, 13) else POINTER(EVP_MD_CTX))
    ]
